File: sorting/merge_sort/merge_sort.py
def merge_sort(elements: list[int]) -> list[int]:

    merge_sort_recursive(elements, 0, len(elements))

    return elements


def merge_sort_recursive(elements: list[int], start: int, end: int) -> list[int]:
    if start >= end - 1:
        return
    middle: int = start + (end - start) // 2
    merge_sort_recursive(elements, start, middle)
    merge_sort_recursive(elements, middle, end)

    merge(elements, start, middle, end)

def merge(elements: list[int], start: int, middle: int, end: int) -> None:
    left_array: list[int] = elements[start:middle]
    right_array: list[int] = elements[middle:end]

    left_size: int = len(left_array)
    right_size: int = len(right_array)

    left_i: int = 0
    right_i: int = 0
    merged_i: int = start

    while left_i < left_size and right_i < right_size:
        if left_array[left_i] < right_array[right_i]:
            elements[merged_i] = left_array[left_i]
            left_i += 1
        else:
            elements[merged_i] = right_array[right_i]
            right_i += 1

        merged_i += 1

    while left_i < left_size:
        elements[merged_i] = left_array[left_i]
        left_i += 1
        merged_i += 1

    while right_i < right_size:
        elements[merged_i] = right_array[right_i]
        right_i += 1
        merged_i += 1


if __name__ == '__main__':
    input: list[int] = [8, 2, 2, 9, 1]
    print(input)
    print(f"{merge_sort(input[:]) = }")

File: sorting/merge_sort/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    public static void main(String[] args) {
        int[] input = { 1, 3, 5, 0, 8, 4, 2 };
        print(input);
        print(mergeSort(input.clone()), "mergeSort");
    }

    static int[] mergeSort(int[] elements) {

        mergeSortRecursive(elements, 0, elements.length);

        return elements;
    }

    static void mergeSortRecursive(int[] elements, int start, int end) {
        if (start >= end - 1)
            return;
        int middle = start + (end - start) / 2;
        mergeSortRecursive(elements, start, middle);
        mergeSortRecursive(elements, middle, end);
        merge(elements, start, middle, end);
    }

    static void merge(int[] elements, int start, int middle, int end) {
        int[] leftArray = Arrays.copyOfRange(elements, start, middle);
        int[] rightArray = Arrays.copyOfRange(elements, middle, end);

        int leftSize = leftArray.length;
        int rightSize = rightArray.length;

        int left_i = 0;
        int right_i = 0;
        int merged_i = start;

        while (left_i < leftSize && right_i < rightSize) {
            if (leftArray[left_i] < rightArray[right_i]) {
                elements[merged_i] = leftArray[left_i];
                left_i++;
            } else {
                elements[merged_i] = rightArray[right_i];
                right_i++;
            }
            merged_i++;
        }

        while (left_i < leftSize) {
            elements[merged_i] = leftArray[left_i];
            left_i++;
            merged_i++;
        }
        while (right_i < rightSize) {
            elements[merged_i] = rightArray[right_i];
            right_i++;
            merged_i++;
        }
    }

    private static void print(int[] array, String text) {
        System.out.print(text + " ");
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    private static void print(int[] array) {
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();

    }
}